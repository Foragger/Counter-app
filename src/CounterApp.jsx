//1. Crear un nuevo componente dentro de la carpeta SRC llamado CounterApp 
//2. El CounterApp debe de ser un Functional Component
//3. El contenido del CounterApp debe de ser: 

// <h1>CounterApp</h1>
// <h2>{value}</h2>

//4. Donde "value" es una propiedad enviada desde el padre hacia el componente
//   CounterApp(Debe ser numerica validada con PropTypes)
//5. Reemplazar en el index.js o main.jsx  el componente de <FirstApp /> por el componente <CounterApp/>
//   (No se olviden del value que debe de ser un numero)
//6. Asegurense de no tener errores ni warnings(Cualquier warning no usado, comentar en el codigo).


import React, { useState } from 'react';
import PropTypes from 'prop-types';



export const CounterApp = ({ value  }) => {

    //la cte counter recibe el valor inicial que le mandamos al useState, que es 10
    
    //la segunda cte es una funcion que hace funcionar al useState y se tiene que llamar a esa para cambiar al counter.
    
    const [counter, setCounter] = useState( value );
    
    //y cuando lo llamamos el setCounter cambio y en consecuencia se tiene que volver a renderizar
    
    const handleAdd = ( ) => {

        // En la version 18 solo se actualiza cuando termina de ejecutarse toda la funcion. En versiones anteriores no. 
        //counter = counter + 1; este no funciona porque counter es una Constante y no se puede cambiar.

        //entonces llamamos a setCounter que es una funcion que esta dentro de useState
        //y no se usa counter ++ porque daria error. porque de nuevo se estaria modificando el counter de manera directa.
        
        //setCounter( counter + 1);
       
        //Tambien no hace falta usar el valor de counter
        //setCounter( 1000 ) ;

        //a su vez tambien no hace falta usar el name implicito de la cte counter
        //tambien usar adentro como si fuese una funcion. 
        setCounter( (c) => c + 1 ) 
    }

    const handleReset = () =>
    {
        setCounter( 0 );
    }

    const handleMinus = () =>
    {
        setCounter( counter - 1 );
    }
    return(
        <>
            <h1>CounterApp</h1>
            <h2> {counter} </h2>
            <button onClick={ handleAdd }>
                +1
            </button>
            <button onClick={ handleMinus }>
                -1
            </button>
            <button onClick={ handleReset }>
                Reset
            </button>
        </>
    )
}

CounterApp.propTypes = {
    value : PropTypes.number.isRequired,
}

