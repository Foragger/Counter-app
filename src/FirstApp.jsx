import React from 'react'
import PropTypes from 'prop-types';
//const newMessage = [0,1,2,3,4,5,6,7,8,9];
// const newMessage = {
//     message: 'Roberto',
//     title: 'Stick',
// }


// const getSaludo = (saludo) => {
//     return (saludo ? 'Hola Mundo' : 'Adios mundo');
// }



// si la variable no necesita ser renderizada, 
//entonces se puede poner fuera de la funcion para que react 
//no lo renderice todo el tiempo
// o mejor dicho si no hay dependencia en la pagina, que necesite estar cambiando, se los coloca fuera. 

//no nos dejara mostar el objeto cuando ponemos: <h1>{newMessage}</h1>
//para eso usamos el JSON.stringify

// las funciones no pueden ser Async porque es un objeto

// es mas facil destructurar el props
//const FirstApp = (props) => {

export const FirstApp = ({ title , subTitle, name }) => {
    

    return (
        <>
            {/* <h1>{getSaludo(true)}</h1>
            <h1>{getSaludo(false)}</h1> */}
            {/* <h1>{newMessage}</h1> */}
            {/* <h1>{JSON.stringify(newMessage)}</h1> */}

            <h1>{ title }</h1>
            {/* <div data-testid="test-title">{ title }</div> */}

            {/* <h2>FirstApp</h2> */}
            {/* <p>Soy tito</p> */}

            <p>{ subTitle }</p>
            <p>{ subTitle }</p>
            <p>{ name }</p>
        </>
    )
}

FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string,
    name : PropTypes.string,
}

FirstApp.defaultProps = {
    title : 'No hay titulo',
    subTitle : 'No hay subtitulo',
    name : 'No hay nombre',
}