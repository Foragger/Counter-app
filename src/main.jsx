import React from 'react';
import ReactDOM from 'react-dom/client';
import './styles.css';
import { CounterApp } from './CounterApp';
import { FirstApp } from './FirstApp';


ReactDOM.createRoot( document.getElementById('root')).render(
    <React.StrictMode>
        {/* <CounterApp value={10}/> */}
        <FirstApp title="hola, soy Tito" />
    </React.StrictMode>
)