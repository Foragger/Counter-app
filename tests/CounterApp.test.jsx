import { CounterApp } from "../src/CounterApp";
import { render,screen } from "@testing-library/react";


// ## test: debe de hacer match con el snapshot
// ## test: debe de mostrar el valor inicial de 100 <CounterApp value={100}>


describe('Haciendo pruebas en CounterApp',() => { 
    const valor = 100;
    test('debe de hacer match con el snapshot',() => { 

        const {container} = render(
            <CounterApp value={valor}/>
        )
        console.log(container);
        expect(container).toMatchSnapshot();

    });
    test('debe de mostrar el valor inicial de 100',() => { 

        const {container} = render(
            <CounterApp value = {valor}/>
        )
        const h2 = container.querySelector('h2');
        expect(h2.innerHTML).toBe(valor.toString());

        // otras formas de hacerlo : 
        // expect(screen.getByText(100)).toBeTruthy();
        // expect(screen.getByRole('heading', { level: 2}).innerHTML ).toContain('100');
    })
});