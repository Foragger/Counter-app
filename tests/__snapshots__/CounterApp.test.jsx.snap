// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Haciendo pruebas en CounterApp debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
    100
  </h2>
  <button>
    +1
  </button>
  <button>
    -1
  </button>
  <button>
    Reset
  </button>
</div>
`;
